// can-observable-array/dist/can-observable-array.js
"use strict";

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance"); }

function _iterableToArray(iter) { if (Symbol.iterator in Object(iter) || Object.prototype.toString.call(iter) === "[object Arguments]") return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = new Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _get(target, property, receiver) { if (typeof Reflect !== "undefined" && Reflect.get) { _get = Reflect.get; } else { _get = function _get(target, property, receiver) { var base = _superPropBase(target, property); if (!base) return; var desc = Object.getOwnPropertyDescriptor(base, property); if (desc.get) { return desc.get.call(receiver); } return desc.value; }; } return _get(target, property, receiver || target); }

function _superPropBase(object, property) { while (!Object.prototype.hasOwnProperty.call(object, property)) { object = _getPrototypeOf(object); if (object === null) break; } return object; }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

var canReflect = require("can-reflect");

var _require = require("can-observable-mixin"),
    makeDefineInstanceKey = _require.makeDefineInstanceKey,
    mixins = _require.mixins,
    mixinMapProps = _require.mixinMapProps,
    mixinTypeEvents = _require.mixinTypeEvents;

var _require2 = require("./helpers"),
    dispatchLengthPatch = _require2.dispatchLengthPatch;

var ProxyArray = require("./proxy-array")();

var queues = require("can-queues");

var type = require("can-type"); // symbols aren't enumerable ... we'd need a version of Object that treats them that way


var localOnPatchesSymbol = "can.patches";
var onKeyValueSymbol = Symbol.for("can.onKeyValue");
var offKeyValueSymbol = Symbol.for("can.offKeyValue");
var metaSymbol = Symbol.for("can.meta");

function convertItem(Constructor, item) {
  if (Constructor.items) {
    var definition = mixins.normalizeTypeDefinition(Constructor.items.type || Constructor.items);
    return canReflect.convert(item, definition);
  }

  return item;
}

function convertItems(Constructor, items) {
  if (items.length) {
    if (Constructor.items) {
      for (var i = 0, len = items.length; i < len; i++) {
        items[i] = convertItem(Constructor, items[i]);
      }
    }
  }
}

var MixedInArray = mixinTypeEvents(mixinMapProps(ProxyArray));

var ObservableArray =
/*#__PURE__*/
function (_MixedInArray) {
  _inherits(ObservableArray, _MixedInArray);

  // TODO define stuff here
  function ObservableArray(items, props) {
    var _this;

    _classCallCheck(this, ObservableArray);

    // Arrays can be passed a length like `new Array(15)`
    var isLengthArg = typeof items === "number";

    if (isLengthArg) {
      _this = _possibleConstructorReturn(this, _getPrototypeOf(ObservableArray).call(this, items));
    } else if (arguments.length > 0 && !Array.isArray(items)) {
      throw new Error("can-observable-array: Unexpected argument: " + _typeof(items));
    } else {
      _this = _possibleConstructorReturn(this, _getPrototypeOf(ObservableArray).call(this));
    }

    mixins.finalizeClass(this instanceof ObservableArray ? this.constructor : void 0);
    mixins.initialize(_assertThisInitialized(_this), props || {});

    for (var i = 0, len = items && items.length; i < len; i++) {
      _this[i] = convertItem(this instanceof ObservableArray ? this.constructor : void 0, items[i]);
    }

    return _possibleConstructorReturn(_this);
  }

  _createClass(ObservableArray, [{
    key: "push",
    value: function push() {
      var _get2;

      for (var _len = arguments.length, items = new Array(_len), _key = 0; _key < _len; _key++) {
        items[_key] = arguments[_key];
      }

      convertItems(this.constructor, items);
      return (_get2 = _get(_getPrototypeOf(ObservableArray.prototype), "push", this)).call.apply(_get2, [this].concat(items));
    }
  }, {
    key: "unshift",
    value: function unshift() {
      var _get3;

      for (var _len2 = arguments.length, items = new Array(_len2), _key2 = 0; _key2 < _len2; _key2++) {
        items[_key2] = arguments[_key2];
      }

      convertItems(this.constructor, items);
      return (_get3 = _get(_getPrototypeOf(ObservableArray.prototype), "unshift", this)).call.apply(_get3, [this].concat(items));
    }
  }, {
    key: "filter",
    value: function filter(callback) {
      if (_typeof(callback) === "object") {
        var props = callback;

        callback = function callback(item) {
          for (var prop in props) {
            if (item[prop] !== props[prop]) {
              return false;
            }
          }

          return true;
        };
      }

      return _get(_getPrototypeOf(ObservableArray.prototype), "filter", this).call(this, callback);
    }
  }, {
    key: "forEach",
    value: function forEach() {
      for (var _len3 = arguments.length, args = new Array(_len3), _key3 = 0; _key3 < _len3; _key3++) {
        args[_key3] = arguments[_key3];
      }

      return Array.prototype.forEach.apply(this, args);
    }
  }, {
    key: "splice",
    value: function splice() {
      for (var _len4 = arguments.length, args = new Array(_len4), _key4 = 0; _key4 < _len4; _key4++) {
        args[_key4] = arguments[_key4];
      }

      var index = args[0],
          howMany = args[1],
          added = [],
          i,
          len,
          listIndex,
          allSame = args.length > 2;
      index = index || 0; // converting the arguments to the right type

      for (i = 0, len = args.length - 2; i < len; i++) {
        listIndex = i + 2; // This should probably be a DefineObject but how?

        args[listIndex] = convertItem(this.constructor, args[listIndex]); //args[listIndex] = this.__type(args[listIndex], listIndex);

        added.push(args[listIndex]); // Now lets check if anything will change

        if (this[i + index] !== args[listIndex]) {
          allSame = false;
        }
      } // if nothing has changed, then return


      if (allSame && this.length <= added.length) {
        return added;
      } // default howMany if not provided


      if (howMany === undefined) {
        howMany = args[1] = this.length - index;
      }

      queues.batch.start();

      var removed = _get(_getPrototypeOf(ObservableArray.prototype), "splice", this).apply(this, args);

      queues.batch.stop();
      return removed;
    }
  }, {
    key: Symbol.for("can.splice"),

    /* Symbols */
    value: function value(index, deleteCount, insert) {
      return this.splice.apply(this, _toConsumableArray([index, deleteCount].concat(insert)));
    }
  }, {
    key: Symbol.for("can.onPatches"),
    value: function value(handler, queue) {
      this[onKeyValueSymbol](localOnPatchesSymbol, handler, queue);
    }
  }, {
    key: Symbol.for("can.offPatches"),
    value: function value(handler, queue) {
      this[offKeyValueSymbol](localOnPatchesSymbol, handler, queue);
    }
  }, {
    key: Symbol.for("can.isListLike"),
    get: function get() {
      return true;
    }
  }], [{
    key: Symbol.for("can.new"),
    value: function value(items) {
      var array = items || [];
      return new this(array);
    }
  }, {
    key: "convertsTo",
    value: function convertsTo(Type) {
      var ConvertedType = type.convert(Type);

      var ArrayType =
      /*#__PURE__*/
      function (_this2) {
        _inherits(ArrayType, _this2);

        function ArrayType() {
          _classCallCheck(this, ArrayType);

          return _possibleConstructorReturn(this, _getPrototypeOf(ArrayType).apply(this, arguments));
        }

        _createClass(ArrayType, null, [{
          key: "items",
          get: function get() {
            return ConvertedType;
          }
        }]);

        return ArrayType;
      }(this);

      var name = "ConvertedObservableArray<".concat(canReflect.getName(Type), ">");
      canReflect.setName(ArrayType, name);
      return ArrayType;
    }
  }, {
    key: Symbol.species,
    get: function get() {
      return this;
    }
  }]);

  return ObservableArray;
}(MixedInArray);

var mutateMethods = {
  "push": function push(arr, args) {
    return [{
      index: arr.length - args.length,
      deleteCount: 0,
      insert: args,
      type: "splice"
    }];
  },
  "pop": function pop(arr) {
    return [{
      index: arr.length,
      deleteCount: 1,
      type: "splice"
    }];
  },
  "shift": function shift() {
    return [{
      index: 0,
      deleteCount: 1,
      type: "splice"
    }];
  },
  "unshift": function unshift(arr, args) {
    return [{
      index: 0,
      deleteCount: 0,
      insert: args,
      type: "splice"
    }];
  },
  "splice": function splice(arr, args) {
    return [{
      index: args[0],
      deleteCount: args[1],
      insert: args.slice(2),
      type: "splice"
    }];
  },
  "sort": function sort(arr) {
    return [{
      index: 0,
      deleteCount: arr.length,
      insert: arr,
      type: "splice"
    }];
  },
  "reverse": function reverse(arr) {
    return [{
      index: 0,
      deleteCount: arr.length,
      insert: arr,
      type: "splice"
    }];
  }
};
canReflect.eachKey(mutateMethods, function (makePatches, prop) {
  var protoFn = ObservableArray.prototype[prop];

  ObservableArray.prototype[prop] = function () {
    var oldLength = this.length; // prevent `length` event from being dispatched by get/set proxy hooks

    this[metaSymbol].preventSideEffects++;
    var result = protoFn.apply(this, arguments);
    this[metaSymbol].preventSideEffects--;
    var patches = makePatches(this, Array.from(arguments));
    dispatchLengthPatch.call(this, prop, patches, this.length, oldLength);
    return result;
  };
});
makeDefineInstanceKey(ObservableArray);
module.exports = ObservableArray;
